spring-integration-core/src/main/java/org/springframework/integration/handler/support/MessagingMethodInvokerHelper.java

```diff
@@ -503,37 +514,46 @@ private void initializeHandler(HandlerMethod candidate) {
 		candidate.initialized = true;
 	}
 
+	@SuppressWarnings("deprecation")
 	private synchronized void initialize() throws Exception {
 		if (!this.initialized) {
-			BeanFactory beanFactory = getBeanFactory();
-			if (beanFactory != null &&
-					beanFactory.containsBean(IntegrationContextUtils.ARGUMENT_RESOLVER_MESSAGE_CONVERTER_BEAN_NAME)) {
+			if (isProvidedMessageHandlerFactoryBean()) {
+				logger.info("Overriding default instance of MessageHandlerMethodFactory with provided one.");
+				this.messageHandlerMethodFactory = getBeanFactory()
+						.getBean(IntegrationContextUtils.MESSAGE_HANDLER_FACTORY_BEAN_NAME, MessageHandlerMethodFactory.class);
+				createHandlerMethod();
+			}
+			else {
+				BeanFactory beanFactory = getBeanFactory();
+				if (beanFactory != null &&
+						beanFactory.containsBean(IntegrationContextUtils.ARGUMENT_RESOLVER_MESSAGE_CONVERTER_BEAN_NAME)) {
 
-				try {
-					MessageConverter messageConverter =
-							beanFactory.getBean(IntegrationContextUtils.ARGUMENT_RESOLVER_MESSAGE_CONVERTER_BEAN_NAME,
-									MessageConverter.class);
+					try {
+						MessageConverter messageConverter =
+								beanFactory.getBean(IntegrationContextUtils.ARGUMENT_RESOLVER_MESSAGE_CONVERTER_BEAN_NAME,
+										MessageConverter.class);
 
-					this.messageHandlerMethodFactory.setMessageConverter(messageConverter);
+						((DefaultMessageHandlerMethodFactory) this.messageHandlerMethodFactory).setMessageConverter(messageConverter);
 
-					HandlerMethodArgumentResolversHolder handlerMethodArgumentResolversHolder =
-							beanFactory.getBean(this.canProcessMessageList
-											? IntegrationContextUtils.LIST_ARGUMENT_RESOLVERS_BEAN_NAME
-											: IntegrationContextUtils.ARGUMENT_RESOLVERS_BEAN_NAME,
-									HandlerMethodArgumentResolversHolder.class);
+						HandlerMethodArgumentResolversHolder handlerMethodArgumentResolversHolder =
+								beanFactory.getBean(this.canProcessMessageList
+												? IntegrationContextUtils.LIST_ARGUMENT_RESOLVERS_BEAN_NAME
+												: IntegrationContextUtils.ARGUMENT_RESOLVERS_BEAN_NAME,
```

Comment:
I'm still not OK do not have an out-of-the-box bean on the matter. And that would simplify a code in here, but at the same time pay attention, please, that we would like to use different argument resolvers for list case and non-list.

With the centralized single `IntegrationContextUtils.MESSAGE_HANDLER_FACTORY_BEAN_NAME` we don't have such a distribution and proper logic here in the in this class.

So, do we need to introduce an  `IntegrationContextUtils.LIST_MESSAGE_HANDLER_FACTORY_BEAN_NAME` as well to cover this situation?

If I'm alone in my "brain swamp", feel free to merge it and let's raise an issue to revise it in the next version!

Reply:
I am not sure i follow. . .
We do not need OOTB bean. As I said in the notes it's _status quo unless_. .  and the _unless_ part is where you decide to override at which point you register a bean of type and name specified. 

Yes, there are quite a lot of additional improvements that can/should be done in this specific class and I've mentioned some of them, but the impact of introducing it now (minor release) may be more then we can/should allow.

Reply:
The problem is that with an explicit `IntegrationContextUtils.MESSAGE_HANDLER_FACTORY_BEAN_NAME` bean in your target application you apply it for all the `MessagingMethodInvokerHelper` instances where some of them really should rely on the `this.canProcessMessageList` and the choice of argument resolves in your bean might be wrong or doesn't fit the list processing logic. I mean that with single global bean you are going to break something in your application without a way to reinstate the proper behavior.

I might agree if you only apply such a global bean for those  `MessagingMethodInvokerHelper`  which are not `this.canProcessMessageList` though...

Reply:
Correct, if I override `MessageHandlerMethodFactory` I am taking full control. I have to take care of everything - as actually shown in test.

Reply:
One more time: we have a special case for the `this.canProcessMessageList`.
See `DefaultConfiguringBeanFactoryPostProcessor.internalArgumentResolversBuilder()` and its logic around `listCapable` parameter.
I'm not sure that a single global bean can address that use-case. And that's definitely might be a case why we have fine-grained it into a separate `IntegrationContextUtils.LIST_ARGUMENT_RESOLVERS_BEAN_NAME`.

Might the case that I just can't find the proper words to explain the issue...

@garyrussell , WDYT?

Reply:
I agree that it would be better to pre-define standard beans for list and not-list case, and allow the user to override either one of them; however, given the timing, and the fact we have to get 10 releases out today, I would tend to go with this compromise (SCSt doesn't need the list-capable invocation - that is generally for aggregation processors).

We can add an issue to the backlog to clean it up properly.

Reply:
OK. Then merging with simple code style polishing.

Reply:
I would also make sure we revisit the lifecycle of `handleMethod` so it's only created when the _actual_ instance of `MessageHandlerMethodFactory` is known.

Now, with regard to _standard beans for list and not-list case_, I still believe it is out of scope for this issue and is the responsibility of the user who configures/overrides `MessageHandlerMethodFactory` to inject and configure the appropriate resolvers. IMHO it must be `all or nothing`. As an example I as a user who provides my own instance of `MessageHandlerMethodFactory` would never expect framework doing anything to it (injecting additional resolvers, etc) other then just using it.

Reply:
Done https://github.com/spring-projects/spring-integration/issues/2690